service_test.go

```diff
@@ -143,427 +144,433 @@ var mockTransactions = []Nip47Transaction{
 }
 var mockTransaction = &mockTransactions[0]
 
-// TODO: split up into individual tests
-func TestHandleEvent(t *testing.T) {
+func TestBackwardsCompatibility(t *testing.T) {
+	// Test without adding a single permission
+}
+
+func TestHandleGetBalanceEvent(t *testing.T) {
 	ctx := context.TODO()
-	svc, _ := createTestService(t)
 	defer os.Remove(testDB)
-	//test not yet receivedEOS
-	res, err := svc.HandleEvent(ctx, &nostr.Event{
-		Kind: NIP_47_REQUEST_KIND,
-	})
-	assert.Nil(t, res)
-	assert.Nil(t, err)
-	//now signal that we are ready to receive events
-	svc.ReceivedEOS = true
-
-	senderPrivkey := nostr.GeneratePrivateKey()
-	senderPubkey, err := nostr.GetPublicKey(senderPrivkey)
+	mockLn, err := NewMockLn()
 	assert.NoError(t, err)
-	//test lnbc.. payload without having an app registered
-	ss, err := nip04.ComputeSharedSecret(svc.cfg.IdentityPubkey, senderPrivkey)
+	svc, err := createTestService(mockLn)
 	assert.NoError(t, err)
-	payload, err := nip04.Encrypt(nip47PayJson, ss)
+	_, app, ss, err := createUserWithApp(svc)
+	assert.NoError(t, err)
+
+	request := &Nip47Request{}
+	err = json.Unmarshal([]byte(nip47GetBalanceJson), request)
 	assert.NoError(t, err)
-	res, err = svc.HandleEvent(ctx, &nostr.Event{
-		ID:      "test_event_1",
+
+	// without permission
+	payload, err := nip04.Encrypt(nip47GetBalanceJson, ss)
+	assert.NoError(t, err)
+	reqEvent := &nostr.Event{
 		Kind:    NIP_47_REQUEST_KIND,
-		PubKey:  senderPubkey,
+		PubKey:  app.NostrPubkey,
 		Content: payload,
-	})
+	}
+
+	reqEvent.ID = "test_get_balance_without_permission"
+	res, err := svc.HandleGetBalanceEvent(ctx, request, reqEvent, *app, ss)
 	assert.NoError(t, err)
 	assert.NotNil(t, res)
-	received := &Nip47Response{}
-	decrypted, err := nip04.Decrypt(res.Content, ss)
+
+	response, err := decryptResponse(res, ss, nil)
 	assert.NoError(t, err)
-	err = json.Unmarshal([]byte(decrypted), received)
+	assert.NotNil(t, response)
+	assert.Equal(t, NIP_47_ERROR_RESTRICTED, response.Error.Code)
+
+	// with permission
+	expiresAt := time.Now().Add(24 * time.Hour)
+	appPermission := &AppPermission{
+		AppId:         app.ID,
+		App:           *app,
+		RequestMethod: NIP_47_GET_BALANCE_METHOD,
+		ExpiresAt:     expiresAt,
+	}
+	err = svc.db.Create(appPermission).Error
+	assert.NoError(t, err)
+
+	reqEvent.ID = "test_get_balance_with_permission"
+	res, err = svc.HandleGetBalanceEvent(ctx, request, reqEvent, *app, ss)
 	assert.NoError(t, err)
-	assert.Equal(t, received.Error.Code, NIP_47_ERROR_UNAUTHORIZED)
```

Comment:
why was this assert removed?

Reply:
we do not have a test for this any more, it would be good to make sure using an unknown pubkey returns unauthorized

Reply:
We might not be able to do that because that part is handled by `HandleEvent` which doesn't return anything

Reply:
I think we can extract the app finding logic to a new method